junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/JupiterTestDescriptor.java

```diff
@@ -88,9 +97,33 @@
 		// @formatter:on
 	}
 
-	protected static <E extends AnnotatedElement> String determineDisplayName(E element,
-			Function<E, String> defaultDisplayNameGenerator) {
+	protected static DisplayNameGenerator getDisplayNameGenerator(Class<?> testClass) {
+		Preconditions.notNull(testClass, "Test class must not be null");
+		// first locate direct or meta-direct present annotations
+		DisplayNameGeneration generation = findAnnotation(testClass, DisplayNameGeneration.class).orElse(null);
+		// next, crawl enclosing classes
+		if (generation == null) {
```

Comment:
Class-based inheritance or enclosed-based relation first?

Reply:
Class-based first since that will also pick up annotations on interfaces as well as local declarations which must always override annotations on an enclosing class.